### PR TITLE
FAQ: Why not use Deployment or StatefulSet?

### LINKED ISSUE #2770: Add a FAQ entry describing when you would use Agones vs. StatefulSets
**Is your feature request related to a problem? Please describe.**
A common question with Agones is to understand when you would use it vs. using StatefulSets. We should add a FAQ entry that answers this question that we can point people to.

### PR DESCRIPTION
**Pull request by markmandel**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup

/kind documentation

> /kind feature
> /kind hotfix

**What this PR does / Why we need it**:

The question of "Can't we use a Deployment or a StatefulSet for game server workloads?" comes up a bunch in face to face conversations - so here's a FAQ entry we can point people at.

**Which issue(s) this PR fixes**:

Closes #2770

**Special notes for your reviewer**:

N/A

### CODE REVIEW

File: site/content/en/docs/FAQ/_index.md [hunk 1]
```diff
@@ -184,6 +184,47 @@ This number could vary depending on the underlying scaling capabilities
...
+Dedicated, authoritative game server workloads are _sometimes stateful_, and while 
+not ordered/unordered, game servers are _prioritised_ for both scale down and allocation for player usage.
+
+Game servers are sometimes stateful, because their state only matters if players are playing on them. If no players 
+are playing on a game server, then it doesn't matter if it gets shut down, or replaced, since nobody will notice. 
+But they are stateful (most often in-memory state, for the game simulation) when players are playing on them, and 
+therefore can't be shutdown while that is going on.
+
+Game Server workloads are also _prioritised_, in that the order of demarkating game servers for player connection 
+and also on game server scale down impact optimial usage of  the hosting infrastructure.
```
Comment (gongmax) [inline site/content/en/docs/FAQ/_index.md:218]: nit: extra space between "of" and "the".

File: site/content/en/docs/FAQ/_index.md [hunk 2]
```diff
@@ -184,6 +184,47 @@ This number could vary depending on the underlying scaling capabilities
 of your cloud provider, Kubernetes cluster configuration, and your GameServer Ready startup time, and
 therefore we recommend you always run your own load tests for your specific game and game server containers.
 
+## Architecture
+
+### Can't we use a Deployment or a StatefulSet for game server workloads?
+
+Kubernetes [Deployments](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) where built for
```
Comment (roberthbailey) [inline site/content/en/docs/FAQ/_index.md:191]: nit: s/where/were

File: site/content/en/docs/FAQ/_index.md [hunk 3]
```diff
@@ -184,6 +184,47 @@ This number could vary depending on the underlying scaling capabilities
...
+
+Kubernetes [Deployments](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) where built for 
+_unordered_, _stateless_ workloads. That is, workloads that are essentially homogeneous 
+between each instance, and therefore it doesn't matter which order they are scaled up, or scaled down.
+
+A set of web servers behind the same load balancer are a perfect example of this. The configuration and application 
+code between instances is the same, and as long as there are enough replicas to handle the requests coming through a 
+load balancer, if we scale from 10 to 5, it doesn't matter which ones are removed and in which order.
+
+Kubernetes [StatefulSets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)
```
Comment (roberthbailey) [inline site/content/en/docs/FAQ/_index.md:199]: nit: s/where/were

File: site/content/en/docs/FAQ/_index.md [hunk 4]
```diff
@@ -184,6 +184,47 @@ This number could vary depending on the underlying scaling capabilities
...
+where built for _ordered_, _stateful_ workloads. That is, workloads in which each instance is 
+essentially heterogeneous, and for reliability and predictability it's extremely important that scale up happens in 
+order (0 ,1, 2, 3) and scaling down happens in reverse (3, 2, 1, 0).
+
+Databases are a great use case for a StatefulSet, since (depending on the database), instance 0 may be the primary, 
+and instances 1, 2, 3+ may be replicas. Knowing that the order of scale up and down in completely reliable to both
```
Comment (roberthbailey) [inline site/content/en/docs/FAQ/_index.md:205]: nit: s/in/is

File: site/content/en/docs/FAQ/_index.md [hunk 5]
```diff
@@ -184,6 +184,47 @@ This number could vary depending on the underlying scaling capabilities
...
+
+Dedicated, authoritative game server workloads are _sometimes stateful_, and while 
+not ordered/unordered, game servers are _prioritised_ for both scale down and allocation for player usage.
+
+Game servers are sometimes stateful, because their state only matters if players are playing on them. If no players 
+are playing on a game server, then it doesn't matter if it gets shut down, or replaced, since nobody will notice. 
+But they are stateful (most often in-memory state, for the game simulation) when players are playing on them, and 
+therefore can't be shutdown while that is going on.
+
+Game Server workloads are also _prioritised_, in that the order of demarkating game servers for player connection
```
Comment (roberthbailey) [inline site/content/en/docs/FAQ/_index.md:217]: nit: s/demarkating/demarcating

File: site/content/en/docs/FAQ/_index.md [hunk 6]
```diff
@@ -184,6 +184,47 @@ This number could vary depending on the underlying scaling capabilities
...
+Game Server workloads are also _prioritised_, in that the order of demarkating game servers for player connection 
+and also on game server scale down impact optimial usage of  the hosting infrastructure.   
+
+For example, in Cloud based workloads, you will want to pack the game servers that have players on them as tightly 
+as possible across as few Nodes as possible, while on scale down, will want to prioritise removing game servers from  
+Nodes that are the most empty to create empty Nodes that then can be deleted - thereby using the least amount of 
+infrastructure as possible.
+  
+So while one might be able to use Deployments and/or StatefulSets to run game server workloads, it will be extremely 
+hard (impossible? 🤔) to run as optimally as a tailored solutions such as Agones.
```
Comment (roberthbailey) [inline site/content/en/docs/FAQ/_index.md:226]: nit: solution (singular)

**Review by zmerlynn — APPROVED** [review]
LGTM once other comments are resolved.

**Comment by markmandel** [comment]
This should be good to go now!

**Review by roberthbailey — APPROVED** [review]

**Pull request merged**